src/qibolab/platforms/platform.py

```diff
@@ -18,7 +18,16 @@ def connect(self):
         self.is_connected = True
 
     def setup(self):
-        self.design.setup(self.qubits, self.channels, **self.options)
+        for qubit in self.qubits.values():
+            if qubit.flux is not None:
+                # set flux offset
+                qubit.flux.offset = qubit.sweetspot
+                # Set flux filters (useful for CZ gates)
+                qubit.flux.filter = {
+                    "feedforward": qubit.ff_filter,
+                    "feedback": qubit.fb_filter,
+                }
```

Comment:
Since all these attributes are already contained in the qubit, instead of actually setting other properties, wouldn't it be possible to make suitable accessors?

Of course this is only true at the `qubit` level, but not at the `flux` level, that only holds this information after assignment.

So, if the `Flux` object is needed independently of a `Qubit` object, the answer is definitely no, and you should keep going.
If `Flux` is just a grouping of `Qubit` attributes, then it might be possible to design suitable accessors, instead of duplicating the information.

Reply:
> I would say `flux` is needed because it holds additional information that does not appear in this cell, for example `ports`, which are used to play flux pulses for doing two-qubit gates. Also it is another `Channel` connected to the qubits so I believe it makes sense to treat it similarly to `readout` and `drive`.

I'm not worried about having a `flux` object, but the problem I see is that if you traverse the *whole structure* of (all the memory pointed by) `Qubit`, you will have redundant information.

In particular, I believe this information not to be temporarily redundant (and thus you are just initializing to one value, but it would be fine to modify one and not the other), but also it should be in principle the same object.
So, if you have the same information replicated, you have to make an extra effort to keep it consistent, otherwise the information you will get might depend on the way you access it, causing possible bugs.

Reply:
> This also brings in my mind another point, which is not fully implemented yet in my driver. Ideally `qubit.flux.offset` `qubit.sweetspot` should be two different things. `qubit.flux.offset` represents a DC voltage that the real instrument is applying, which is variable and tunable by the user. `qubit.sweetspot` is the value of this voltage that maximizes the qubit's frequency, which is a physical property measured during calibration and almost constant (modulo fluctuations in temperature). Qubits are generally operated at their sweetspots, which is what line 24 is ensuring, however for other experiments someone might want to tune `qubit.flux.offset` seperately. For this we will need to implement a proper setter that updates the corresponding value in the instrument. Regardless, I think it makes sense to keep the two seperated.

This is fine: in case you want to keep the default value of `offset` to `sweetspot`, you can write a suitable accessor and mutator:
```py
@dataclass
class Flux:
    _offset: Optional[float]
    sweetspot: float

    @property
    def offset(self):
        if self._offset is None:
            return self.sweetspot

    @offset.setter
    def offset(self, value):
        self._offset = value
```

It is such a common pattern that I wonder if there is an even more concise way of implementing it...
Maybe with [`field`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field)?

Reply:
> I could move the `offset` to `Qubit` because at least in our current setup each qubit has its own fluxline, but I am not sure if that's the case for all chips. Also I think the way it is now makes more sense because the `offset` is really a property of the `Channel` while `sweetspot` is a property of the `Qubit`.

No, you should definitely keep doing what you are. We should make sensible abstractions, thus matching the conceptual relations in the first place, and membership is one of the most relevant.
We will rearrange the structure given the conceptual constraints.

However, at the moment I'm missing the bigger picture, so keep as it is and we will reconsider the overall structure in a while.

Reply:
> Btw, similar arguments can be said for all characterization attributes of `Qubit`. For example there is `qubit.readout_frequency` that is not used in the code, because the frequencies we are playing are given in the pulse objects. If you want to measure the qubit you need to send the pulse at its readout frequency, but obviously the driver allows playing any frequency.

I see your point, and defaults are less worrisome than matching values, since they are only equal at initialization time. However, if it is possible also to redirect with a `None` and a getter, this would be even better, because you would know exactly where the default value is coming from, without hunting for the assignment (if you are getting the default value, then you would pass through the getter in the debugger, while now, with all the assignments, you risk to loose track about where things are set).

In any case, I would not block the PR on this issue alone, so feel free to mark as solved. We will reconsider at a later stage, together with the former one. One of the good features of the current factoring should be simplifying the code structure, then it should be easier to reason about it afterward.

Reply:
> I see your point, and defaults are less worrisome than matching values, since they are only equal at initialization time. However, if it is possible also to redirect with a `None` and a getter, this would be even better, because you would know exactly where the default value is coming from, without hunting for the assignment (if you are getting the default value, then you would pass through the getter in the debugger, while now, with all the assignments, you risk to loose track about where things are set).

Thanks @AleCandido. Have a look at the last commit where I do this with setters. It is a few more lines of code but as you say it may be easier to keep track.

Regarding the `Channel` object, as I mentioned before, my idea is to use it to replace some of the dictionaries that appear in the `MultiqubitPlatform` and used in the qblox driver, specifically the ones that appear here https://github.com/qiboteam/qibolab/blob/448a54f2685dbafb4908473c8c3eb3a22ccc89c0/src/qibolab/platforms/multiqubit.py#L48 and beyond. 

The current `Channel` object seems sufficient for the QM driver, but it is not clear if it will also work for qblox and other platforms. I guess when we start implementing/refactoring these we can adjust it accordingly. Considering the current priorities I would leave this exercise for a later stage.

Reply:
The solution looks almost perfect, but for two considerations:
- a potential one: if a `Channel` might have more than one `qubit` (and we already know), maybe it is not ideal to assign one just for the time being (but this I don't know at all)
- an abstract one: self-referential structures might be cumbersome, `qubit -> channel["flux"] -> qubit`, but especially when you are responsible for memory management on your own, that is not the case of Python; however, sometimes not having a membership tree (because there are loops) might result in counterintuitive behavior, but this we'll face the moment we get there

I know that you tried to follow my suggestions, and I like more the current status, but I'm just being honest: I don't have an optimal solution, and this is a problem I frequently encountered (how to read a property of the parent from a member element, and passing down values every time, copying them, and register a reference have all their own drawbacks).
However, this is especially frequent with graph-like structures (where you have nodes and links, but it is not clear that one of the two is "more important" than the other). One possible solution is to downgrade references to labels in a global storage, at least they don't mess up memory. But this is not the case of Python, so let's keep going as it is right now, and just reconsider afterward.

Reply:
> * a potential one: if a `Channel` might have more than one `qubit` (and we already know), maybe it is not ideal to assign one just for the time being (but this I don't know at all)

Since you mention this, yes a `Channel` can have more than one qubit, particularly:

* readout channels certainly have more than one qubit (in the current setup we have two readout channels for five qubits),
* drive and flux have only one qubit,

however I have no idea if this is true for all possible quantum computer setups. If I had to bet I would say no. Here I am only assigning the `channel.qubit` for the case of `flux` where I know that is only one qubit, at least in our setup. 

Note that here the `Channel` object is not optimal because it tries to satisfy many different roles. The cleaner way would be to have a different object for each channel role (eg. `Readout`, `Drive`, `Flux`, etc.), because their attributes would also differ. For example flux channels do not have local oscillators as they only send DC signals (which again I am not sure if it will be true for other setups). The only reason I went with a single object here was because I did not want to overpopulate the code with many different classes before I am sure that they are useful (and also reviewed and agreed by someone other than me!)

Reply:
> The only reason I went with a single object here was because I did not want to overpopulate the code with many different classes before I am sure that they are useful (and also reviewed and agreed by someone other than me!)

I believe it is a perfect strategy. The moment it will be proved to be useful, than I would keep going with the split, but not immediately!
For me, you can resolve the conversation!

Reply:
> however I have no idea if this is true for all possible quantum computer setups. If I had to bet I would say no. Here I am only assigning the channel.qubit for the case of flux where I know that is only one qubit, at least in our setup.

Lets keep this choice for the time being. I am sure changes will be much simpler after these changes.